### ISSUE TITLE
EndBASIC 0.11.0 - exp operator overflow

### ISSUE THREAD

**Issue by steve64** [endbasic/endbasic#244]
I just tried new 0.11.0 and found this issue:

x = 2^31

reports an overflow error

**Comment by jmmv** [comment]
Yes, this is mentioned in the release notes:

*   Reverted integer-only operations to *not* perform automatic upgrades to
    doubles on underflows, overflows, and other error conditions.  This behavior
    was previously added in 0.10.0 but has become problematic with the new type
    propagation during compilation as it pushes too much responsibility to the
    runtime.

Keeping the previous behavior was quite difficult with the current internals, but I had to do these changes to gain performance. The way numbers are handled may need reconsideration though.

For now, forcing the operation to act as a float will work. Try `2.0^31`.

**Comment by steve64** [comment]
Thanks for the quick feedback. I like anyway the separation between integers and floats in EndBASIC (more generally the strong typing with no mixed-type expressions).

**Comment by jmmv** [comment]
OK great then.

Well, note that EndBASIC continues to offer mixed-type expressions by automatically promoting integers to doubles *at compilation time* when the compiler can know that the target of the expression is a double. The problem in `2^31` on its own is that this is an integer expression without extra context, and the runtime cannot change its type when it detects overflow.

An interesting corner case that arises is:

```basic
x = 0.0  ' x is now a double.
x = 2^31 ' Still yields overflow.
```

Even though `x` is defined as a double in the above, the `2^31` subexpression is still evaluated in an integer-only context, so this breaks too. Maybe it shouldn't, but it feels magical to fix it so I think the current stricter behavior feels better.

**Comment by steve64** [comment]
Hi Julio,
I just run the Windows version of EndBASIC 0.11.0 and I notice that characters on the REPL
are duplicated. Any idea about his strange behavior?
Regards,
Stefano

[cid:0d0621ca-f929-4e57-94b8-c54c7be28339]
________________________________
Da: Julio Merino ***@***.***>
Inviato: sabato 20 luglio 2024 16:41
A: endbasic/endbasic ***@***.***>
Cc: steve64 ***@***.***>; Author ***@***.***>
Oggetto: Re: [endbasic/endbasic] EndBASIC 0.11.0 - exp operator overflow (Issue #244)


Closed #244<https://github.com/endbasic/endbasic/issues/244> as completed.
